jxl/src/frame/modular/transforms/palette.rs

```diff
@@ -2,14 +2,244 @@
 //
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
-use crate::{error::Result, frame::modular::ModularBufferInfo};
+use crate::{
+    error::Result,
+    frame::modular::{ModularBufferInfo, ModularGridKind, Predictor},
+    image::{Image, ImageRect},
+};
 
 use super::TransformStep;
 
+const MAX_PALETTE_LOOKUP_TABLE_SIZE: usize = 1 << 16;
+
+const RGB_CHANNELS: usize = 3;
+
+// 5x5x5 color cube for the larger cube.
+const LARGE_CUBE: usize = 5;
+
+// Smaller interleaved color cube to fill the holes of the larger cube.
+const SMALL_CUBE: usize = 4;
+const SMALL_CUBE_BITS: usize = 2;
+// SMALL_CUBE ** 3
+const LARGE_CUBE_OFFSET: usize = SMALL_CUBE * SMALL_CUBE * SMALL_CUBE;
+const IMPLICIT_PALETTE_SIZE: usize = LARGE_CUBE_OFFSET + LARGE_CUBE * LARGE_CUBE * LARGE_CUBE;
+
+fn scale<const DENOM: usize>(value: usize, bit_depth: usize) -> i32 {
+    // return (value * ((1 << bit_depth) - 1)) / denom;
+    // We only call this function with SMALL_CUBE or LARGE_CUBE - 1 as denom,
+    // allowing us to avoid a division here.
+    const {
+        assert!(DENOM == 4, "denom must be 4");
```

Comment:
I'm still not convinced we need the const generic here :-) 

(also, stale comment)

Reply:
Is it? Do you mean because of the case of `DENOM`?

Reply:
Eh, it explains why we `>> 2` to start with. I think I would rather keep it.

Reply:
(I have updated it to at least use the same naming convention as the actual variable.)